src/components/vm/nics/nicBody.tsx

```diff
@@ -271,3 +304,364 @@ export const NetworkTypeAndSourceRow = ({
+        text += _("all");
+    for (let i = 0; i < pf.range.length; i++) {
+        const r = pf.range[i];
+        if (i > 0)
+            text += ",";
+        if (r.exclude == "yes")
+            text += _(" excluding ");
+        text += r.start;
+        if (r.end)
+            text += "-" + r.end;
+        if (!r.exclude)
+            text += " → " + (r.to || r.start) + (r.end ? "-" + (Number(r.to || r.start) + Number(r.end) - Number(r.start)).toString() : "");
+    }
+    text += "/" + pf.proto;
+
+    return text;
+}
+
+export function interfacePortForwardsToDialog(portForwards: VMInterfacePortForward[]): DialogPortForward[] {
+    return portForwards.map(pf => {
+        if (!pf.dev && pf.range.length == 1 && pf.range[0].exclude != "yes") {
+            const r = pf.range[0];
+            return {
+                kind: "simple",
+                address: pf.address || "",
+                proto: pf.proto || "",
+                host: r.start + (r.end ? "-" + r.end : ""),
+                guest: r.to || "",
+            };
+        } else {
+            return {
+                kind: "complex",
+                config: pf,
+            };
+        }
+    });
+}
+
+export function dialogPortForwardsToInterface(portForwards: DialogPortForward[]): VMInterfacePortForward[] {
+    // NOTE: We have to set every field here, none of them can be
+    // "null". Our return value is passed to virt-xml, and we want
+    // virt-xml to set everything and not leave anything behind from
+    // the previous portForward XML element.
+    return portForwards.map((pf: DialogPortForward) => {
+        if (pf.kind == "simple") {
+            const range = pf.host.split("-");
+            return {
+                address: pf.address,
+                dev: "",
+                proto: pf.proto,
+                range: [
+                    {
+                        start: range[0],
+                        end: range[1] || "",
+                        to: pf.guest,
+                        exclude: "",
+                    }
+                ],
+            };
+        } else {
+            return pf.config;
+        }
+    });
+}
+
+function validateAddress(addr: string): string | undefined {
+    if (addr && !ipaddr.isValid(addr))
+        return _("Invalid IP address");
+    return undefined;
+}
+
+function validatePort(port: string): string | undefined {
+    const p = Number(port);
+    if (isNaN(p))
+        return _("Port must be a number.");
+    if (p <= 0 || p > 65535)
+        return _("Port must be 1 to 65535.");
+    return undefined;
+}
+
+function validateHostPort(port: string): string | undefined {
+    if (!port)
+        return _("Host port can not be empty.");
+    const parts = port.split("-");
+    return validatePort(parts[0]) || (parts[1] && validatePort(parts[1]));
+}
+
+function validateGuestPort(port: string): string | undefined {
+    if (port)
+        return validatePort(port);
+    return undefined;
```

Comment:
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2331-790d4f2a-20251111-065830-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)